### ISSUE TITLE
getri change the number of openmp threads

### ISSUE THREAD

**Issue by ye-luo** [xianyi/OpenBLAS#3940]
with 0.3.21 release built by spack
```
'make' '-j8' 'CC=/soft/spack/lib/spack/env/gcc/gcc' 'FC=/soft/spack/lib/spack/env/gcc/gfortran' 'MAKE_NB_JOBS=0' 'ARCH=arm64' 'DYNAMIC_ARCH=1' 'DYNAMIC_OLDER=1' 'TARGET=GENERIC' 'USE_LOCKING=1' 'USE_OPENMP=1' 'USE_THREAD=1' 'RANLIB=ranlib' 'libs' 'netlib' 'shared'
```
Here is the user code.
```
 inline void InvertLU(int n, double* restrict a, int n0, int* restrict piv, double* restrict work, int n1)
 {
   int status;
+  std::cout << "direct size " << n << " before dgetri " << omp_get_max_threads() << std::endl;
   dgetri(n, a, n0, piv, work, n1, status);
+  std::cout << "direct after dgetri " << omp_get_max_threads() << std::endl;
 }
```
and I got output
```
direct size 2 before dgetri 16
direct after dgetri 8
```
The matrix test is 2x2 because this is a low level test.
Changing the number of threads inside dgetri should be fine but the max_threads should be restored as it was upon exiting dgetri.

It seems that this only affects ARM OrangePi A76/55 and and apple M1.
x86 was fine.

**Comment by martin-frbg** [comment]
I cannot reproduce this with the information provided, and there are no code differences at that level of detail (in OpenBLAS at least, also OpenBLAS takes an all-or-nothing approach, either single-threaded or all it can get, but no intermediate thread counts unless someone/something is setting OPENBLAS_NUM_THREADS outside). There were/are some potential issues with querying OpenMP threads in connection with the OMP_PLACES setting, I wonder if you could retest with a build of the current `develop` branch of OpenBLAS ?

**Comment by brada4** [comment]
Can reproduce easily.
If you build openblas without specifying number of threads it takes build system number of processors, then 
driver/others/blas_server_omp.c -   omp_set_num_threads(blas_cpu_number);

Fix is to add NUM_THREADS=256 or something like that to build command or patch Makefile.rule

**Comment by martin-frbg** [comment]
Ah yes you're right, that corner case that is not supposed to happen... would certainly explain why it happens only on a mixed bunch of hosts for him. Using omp_set_num_threads is a bit evil as we have no chance to change it back before returning, there may be better ways to resolve this now.

**Comment by ye-luo** [comment]
My Orange Pi has 8 cores and in an OMP_NUM_THREADS=16  run getri shrinks the thread count to 8 upon exit.
I call getrf first to solve the LU. It doesn't seem to mess up thread count but getri does.

Now I can reproduce the issue on my x86 16 cores (32HT) workstation.
OMP_NUM_THREADS=64. I saw thread count capped as 32.
Keep increasing OMP_NUM_THREADS, fine until OMP_NUM_THREADS=71 when I start to see.
```
OpenBLAS warning: precompiled NUM_THREADS exceeded, adding auxiliary array for thread metadata.
```

> Using omp_set_num_threads is a bit evil as we have no chance to change it back before returning, there may be better ways to resolve this now.

```
const int saved_thread_count = omp_get_max_threads();
getri()
omp_set_num_threads(saved_thread_count);
```

Need a at least a warning and you may choose to print it once.
```
if(saved_thread_count > openblas_can_support) issue a warning to users.
```
Or you can dynamically grow NUM_THREADS during run.

Having dependency on the spec of the build node is not good.

**Comment by martin-frbg** [comment]
Indeed it would need to be checked if dynamically growing the management structure (the "adding auxiliary array" message you saw, a recently added feature) would be possible and sufficient in the OpenMP case. The dependency on the build node is a remnant from the early times when it was expected that one built the library on and for the host where one intended to use it - it _does_ get printed at the end of the build, but of course that does not help with prebuilt binaries obtained e.g. as part of some Linux distribution.

**Comment by brada4** [comment]
There is no performance benefit in exceeding real cores your system has. 
spack built openblas for exactly 8 cores, if you build with more threads package becomes more "universal" as to work on bigger systems too, and supports your experiments oversubscribing your CPU cores. What you hit is a code that should not be reached under normal conditions.

**Comment by martin-frbg** [comment]
Two issues here - oversubscribing the cores is _probably_ not useful, but in any case OpenBLAS should not be silently changing a setting in the code that called it. (Though for this corner case, it is preferable to the trivial alternative, which would be a segfault)

**Comment by brada4** [comment]
spack should expose NUM_THREADS, otherwise BIGNUMA is unreachable for all versions they supply. 
Independently of future improvements in OpenBLAS ....

**Comment by prckent** [comment]
I would like to echo martin-frbg's point. Not only should OpenBLAS not be silently changing the setting in the calling code -- at least a warning or abort should occur -- but it can break the logic of the calling code causing it to generate incorrect results. This should simply never happen.

**Comment by martin-frbg** [comment]
Changing this in the code is likely not trivial - IIRC there are around ten static arrays dimensioned according to the build-time NUM_THREADS argument (or builld host core count in the absence of such argument) and I am not entirely sure if my kludgy allocating of a spillover array (as seen used in the non-OpenMP case) can be applied everywhere. Printing a message from deep down in a software stack does not ensure that it ever gets seen, while aborting would be sure to catch attention in an unfavorable way. Maybe the importance of setting NUM_THREADS could/should be made clearer in the sparse documentation we have - but there is no guarantee that it is read, especially with a precompiled library preinstalled by some distribution. Perhaps the compile-time limit can be communicated internally without using omp_set_num_threads to everybody's surprise, but I'll need some time to look into that.

**Comment by brada4** [comment]
You can use arithmetic for sizing num_threads in most OpenMP implementations without global change.

**Comment by martin-frbg** [comment]
It's starting to look as if the solution is much simpler, as the single call to `omp_set_num_threads()` in blas_server_omp.c appears to be entirely spurious, not in any way required for OpenBLAS' own operation. (Though of course this may not have been the case in the early OpenBLAS from twelve years ago when that line was introduced as part of a fix for #14 )

**Comment by brada4** [comment]
rest of that fix looks rewritten couple of times since, i would drop warning that its sub-optimal if somebody reads those?
